public/app/features/dashboard/api/dashboard_api.ts

```diff
@@ -1,147 +1,50 @@
+import { DashboardDTO } from 'app/types';
+
+import { LegacyDashboardAPI } from './legacy';
+import { DashboardAPI, DashboardWithAccessInfo } from './types';
+import { getDashboardsApiVersion } from './utils';
+import { K8sDashboardAPI } from './v0';
+import { K8sDashboardV2APIStub } from './v2';
+
+// Describes the dashboard DTO types per API version
+export interface ApiVersionDTO {
+  legacy: DashboardDTO;
+  v0: DashboardDTO;
+  // v1: DashboardDTO; TODO[schema]: enable v1 when available
+  v2: DashboardWithAccessInfo<DashboardV2Spec>;
 }
 
-// Implemented using /api/dashboards/*
-class LegacyDashboardAPI implements DashboardAPI {
-  constructor() {}
-
-  saveDashboard(options: SaveDashboardCommand): Promise<SaveDashboardResponseDTO> {
-    dashboardWatcher.ignoreNextSave();
+type DashboardAPIClients = Record<
+  keyof ApiVersionDTO,
+  DashboardAPI<DashboardDTO | DashboardWithAccessInfo<DashboardV2Spec>>
+>;
 
-    return getBackendSrv().post<SaveDashboardResponseDTO>('/api/dashboards/db/', {
-      dashboard: options.dashboard,
-      message: options.message ?? '',
-      overwrite: options.overwrite ?? false,
-      folderUid: options.folderUid,
-    });
-  }
+let clients: Partial<DashboardAPIClients> | undefined;
 
-  deleteDashboard(uid: string, showSuccessAlert: boolean): Promise<DeleteDashboardResponse> {
-    return getBackendSrv().delete<DeleteDashboardResponse>(`/api/dashboards/uid/${uid}`, { showSuccessAlert });
-  }
-
-  getDashboardDTO(uid: string, params?: UrlQueryMap): Promise<DashboardDTO> {
-    return getBackendSrv().get<DashboardDTO>(`/api/dashboards/uid/${uid}`, params);
+export function setDashboardAPI(override: Partial<DashboardAPIClients> | undefined) {
+  if (process.env.NODE_ENV !== 'test') {
+    throw new Error('dashboardAPI can be only overridden in test environment');
   }
+  clients = override;
 }
 
-interface DashboardWithAccessInfo extends Resource<DashboardDataDTO, 'DashboardWithAccessInfo'> {
-  access: Object; // TODO...
-}
-
-// Implemented using /apis/dashboard.grafana.app/*
-class K8sDashboardAPI implements DashboardAPI {
-  private client: ResourceClient<DashboardDataDTO>;
-
-  constructor() {
-    this.client = new ScopedResourceClient<DashboardDataDTO>({
-      group: 'dashboard.grafana.app',
-      version: 'v0alpha1',
-      resource: 'dashboards',
-    });
-  }
-
-  saveDashboard(options: SaveDashboardCommand): Promise<SaveDashboardResponseDTO> {
-    const dashboard = options.dashboard as DashboardDataDTO; // type for the uid property
-    const obj: ResourceForCreate<DashboardDataDTO> = {
-      metadata: {
-        ...options?.k8s,
-      },
-      spec: {
-        ...dashboard,
-      },
+export function getDashboardAPI(): DashboardAPI<DashboardDTO | DashboardWithAccessInfo<DashboardV2Spec>> {
+  if (!clients) {
+    clients = {
+      legacy: new LegacyDashboardAPI(),
+      v0: new K8sDashboardAPI(),
```

Comment:
will v1 be K8s too?

Reply:
@harisrozajac - yes. V0, V1 and V2 already exist (see https://github.com/grafana/grafana/pull/96415).

V0 is k8s api with untyped current dashboard model.
V1 will bring more type safety to the V0 api.
V2 will be the API for serving the v2 schema models.

All of the 3 above are based on k8s architecture.